roboteam_ai/src/coach/OffensiveCoach.cpp

```diff
@@ -0,0 +1,250 @@
+    bool tooClose = false;
+    for (auto &robotPosition : robotPositions) {
+        if (robotPosition.first != self) {
+            if ((position.position - robotPosition.second.position).length() < Constants::ATTACKER_DISTANCE()) {
+                tooClose = true;
+                continue;
+            }
+        }
+    }
+
+    return tooClose;
+}
+
+void OffensiveCoach::calculateNewPositions() {
+    recalculateOffensivePositions();
+
+    roboteam_msgs::GeometryFieldSize field = Field::get_field();
+    double xStart = 0 + marginFromLines;
+    double xEnd = field.right_penalty_line.begin.x - marginFromLines;
+    double yStart = -field.field_width + marginFromLines;
+    double yEnd = field.field_width - marginFromLines;
+
+    int attempts = 30;
+
+    for (int attempt = 0; attempt < attempts; attempt++) {
+        OffensivePosition position = calculateRandomPosition(xStart, xEnd, yStart, yEnd);
+        if (positionTooCloseToRobotPositions(position)) continue;
+        position.score =  CoachHeuristics::calculatePositionScore(position.position);
+
+        if (offensivePositions.empty()) {
+            offensivePositions.push_back(position);
+            continue;
+        }
+
+        compareToCurrentPositions(position);
+    }
+
+    std::sort(offensivePositions.begin(), offensivePositions.end(), compareByScore);
+    if (offensivePositions.size() > maxPositions) {
+        offensivePositions.erase(offensivePositions.begin() + maxPositions, offensivePositions.end());
+    }
+    drawOffensivePoints();
+}
+
+void OffensiveCoach::compareToCurrentPositions(const OffensiveCoach::OffensivePosition &position) {
+    OffensivePosition bestPos = position;
+    auto it = offensivePositions.begin();
+    while (it < offensivePositions.end()) {
+        if ((position.position - it->position).length() < Constants::OFFENSIVE_POSITION_DISTANCE()) {//Constants::ATTACKER_DISTANCE()) {
+            if (it->score >= bestPos.score) {
+                bestPos.position = it->position;
+                bestPos.score = it->score;
+            }
+            offensivePositions.erase(it);
+        }
+        it++;
+    }
+    offensivePositions.push_back(bestPos);
+}
+
+bool OffensiveCoach::compareByScore(OffensivePosition position1, OffensivePosition position2) {
+    return position1.score > position2.score;
+}
+
+Vector2 OffensiveCoach::calculatePositionForRobot(std::shared_ptr<roboteam_msgs::WorldRobot> robot) {
+    if ((robotPositions.find(robot->id) == robotPositions.end())) { // not there yet
+        return getClosestOffensivePosition(robot);
+
+    } else {
+        calculateNewRobotPositions(robot);
+        if ((robotPositions.find(robot->id) == robotPositions.end())) { //not in there
+            return calculatePositionForRobot(robot);
+        } else {
+            return robotPositions[robot->id].position;
+        }
+    }
+}
+
+Vector2 OffensiveCoach::getClosestOffensivePosition(const shared_ptr<roboteam_msgs::WorldRobot> &robot) {
```

Comment:
quite sure you can use hungarian which kinda does the same. Also, you could just simplify this a lot by comparing `robot->pos.dist(position.position)` for each robot.